62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+		return t;
+	}
+
+	// (only works for monotonic curves)
+	static float64_t2 getCurveRoot(double p0, double p1, double p2)
+	{
+		// https://people.csail.mit.edu/bkph/articles/Quadratics.pdf
+		double a = p0 - 2.0 * p1 + p2;
+		double b = 2.0 * (p1 - p0);
+		double c = p0;
+
+		// Solve linear equation
+		if (a == 0.0) return float64_t2(-c / b);
+
+		double det = sqrt(b * b - 4 * a * c);
+		if (b >= 0.0) {
+			return float64_t2(
+				(- b - det) / (2 * a),
+				(2 * c) / (-b - det)
+			);
+		}
+		else
+		{
+			return float64_t2(
+				(2 * c) / (-b + det),
+				(-b + det) / (2 * a)
+			);
+		}
+		//double det = b * b - 4 * a * c;
+		//double rcp = 0.5 / a;
+		//
+		//double detSqrt = sqrt(det) * rcp;
+		//double brcp = b * rcp;
+		//
+		//// Solve linear equation
+		//if (isinf(brcp)) return float64_t2(-c / b);
+		//
+		//return float64_t2(-detSqrt, detSqrt) - brcp;
+	}
+};
+
+Hatch::QuadraticBezier Hatch::Segment::splitCurveRange(QuadraticBezier curve, double left, double right) const
+{
+    return curve.splitCurveTakeLower(right).splitCurveTakeUpper(left / right);
+}
+
+Hatch::QuadraticBezier Hatch::Segment::getSplitCurve() const
+{
+    return splitCurveRange(*originalBezier, t_start, t_end);
+}
+
+bool Hatch::Segment::isStraightLineConstantMajor() const
+{
+	auto major = (uint)SelectedMajorAxis;
+	return originalBezier->p[0][major] == originalBezier->p[1][major] &&
+		originalBezier->p[0][major] == originalBezier->p[2][major];
+}
+
+std::array<double, 2> Hatch::Segment::intersect(const Segment& other) const
+{
+	auto major = (int)SelectedMajorAxis;
+	auto minor = 1-major;
+	// Use line intersections if one or both of the beziers are linear (a = 0)
+	if (originalBezier->isLineSegment())
+	{
+		if (other.originalBezier->isLineSegment())
+		{
+			// Line/line intersection
+			// TODO: make this less shit
+			auto x1 = originalBezier->p[0][minor], y1 = originalBezier->p[0][major],
+				x2 = originalBezier->p[2][minor], y2 = originalBezier->p[2][major],
+				x3 = other.originalBezier->p[0][minor], y3 = other.originalBezier->p[0][major],
+				x4 = other.originalBezier->p[2][minor], y4 = other.originalBezier->p[2][major];
+			auto nx = (x1 * y2 - y1 * x2) * (x3 - x4) - (x1 - x2) * (x3 * y4 - y3 * x4), 
+				ny = (x1 * y2 - y1 * x2) * (y3 - y4) - (y1 - y2) * (x3 * y4 - y3 * x4),
+				d = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4);
+			auto intersectionX = nx / d, intersectionY = ny / d;
+
+			// Return if point is on the lines
+			if (d != 0.0 && 
+				std::min(x1, x2) <= intersectionX && y1 <= intersectionY && std::max(x1, x2) >= intersectionX && y2 >= intersectionY &&
+				std::min(x3, x4) <= intersectionX && y3 <= intersectionY && std::max(x3, x4) >= intersectionX && y4 >= intersectionY)
+				return { other.originalBezier->intersectOrtho(ny / d, major), core::nan<double>() };
+
+			return { core::nan<double>(), core::nan<double>() };
+		}
+
+		// Line/curve intersection
+		// TODO: make this less shit as well
```

Comment:
step1: intersecting two curves/line shouldn't really depend on your choice of minor and major axis, so you shouldn't care about that. maybe only at the end.

step2: construct a `float64_t2x2` for the rotation, and do `mul(rotation, p) `

step3: you don't need to atan2 and then cos sin for constructing a rotation matrix when you already have the direction vector:
```c++
float64_t2  dir = normalize(originalBezier->p[2] - originalBezier->p[0]);
float64_t2x2 rotation = { {dir.x, -dir.y}, {dir.y, dir.x}};
QuadraticBezier rotatedOther = {
    mul(rotation, other.originalBezier->p[0] - originalBezier->p[0]),
    mul(rotation, other.originalBezier->p[1] - originalBezier->p[0]),
    mul(rotation, other.originalBezier->p[2] - originalBezier->p[0])
};
```

step4:

Reply:
for hint and similar work you can see my ellipse subdivision function in the end of curves.cpp